### PR TITLE
Support Transport and TransportWithMessageCredential security mode

### PR DESCRIPTION
**Pull request by ZhaodongTian**
Support TransportWithMessageCredential security mode : #3969 
Fixed issues: #4412 and #4498

### CODE REVIEW

**Comment by HongGit** [comment]
@ZhaodongTian can you please describe this change so we understand why these changes are needed for dotnet-svcutil to support TransportWithMessageCredential feature.

File: src/dotnet-svcutil/lib/src/FrameworkFork/System.ServiceModel/System/ServiceModel/FederatedMessageSecurityOverHttp.cs [hunk 1]
```diff
@@ -139,7 +139,93 @@ public Collection<XmlElement> TokenRequestParameters
 
         internal static bool TryCreate(SecurityBindingElement sbe, bool isSecureTransportMode, bool isReliableSession, MessageSecurityVersion version, out FederatedMessageSecurityOverHttp messageSecurity)
         {
-            throw new NotImplementedException();
+            Fx.Assert(null != sbe, string.Empty);
```
[the remark this answers is not in the document]
Comment (ZhaodongTian) [reply src/dotnet-svcutil/lib/src/FrameworkFork/System.ServiceModel/System/ServiceModel/FederatedMessageSecurityOverHttp.cs:142]: Any bindings with Security mode will have to call [WSFederationHttpBinding.TryCreate](https://github.com/dotnet/wcf/blob/9bf1eea452c8bfad9a4f833097ac3d4c093d987e/src/dotnet-svcutil/lib/src/FrameworkFork/System.ServiceModel/System/ServiceModel/WSHttpBindingBase.cs#L336) method . The implementation for this is forked from framework. Do you think we can update code here to remove Federation related?

**Comment by ZhaodongTian** [comment]
> Does anything need to be done to support WSHttpBinding and WS2007HttpBinding? I'm not seeing tests added for these scenarios either. I believe the existing tests download the WSDL from some of the WCF test endpoints to validate.

I have updated code to support them, i will summit the changes together with test case

**Comment by HongGit** [comment]
@ZhaodongTian or @imcarolwang could you please rebase this PR?

File: src/dotnet-svcutil/lib/src/FrameworkFork/System.ServiceModel/System/ServiceModel/FederatedMessageSecurityOverHttp.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mconnew) [reply src/dotnet-svcutil/lib/src/FrameworkFork/System.ServiceModel/System/ServiceModel/FederatedMessageSecurityOverHttp.cs:142]: Yes, remove the call to WSFederationHttpBinding.TryCreate

**Comment by mconnew** [comment]
Other than needing to remove call to the federation binding to create that from the WSDL,

**Comment by mconnew** [comment]
@HongGit, I realized there's another problem caused by the recent security tightening of our test services. The tests for svcutil all hit wcfcoresvr5 which means they can't be run from a remote dev machine any more. We should modify these tests to use localhost unless there's a ServiceUri set, like the WCF tests do.

File: src/dotnet-svcutil/lib/src/CodeDomFixup/EndpointSelector.cs [hunk 2]
```diff
@@ -62,6 +62,10 @@ private static bool IsBindingSupported(Binding binding)
                     {
                         s_bindingValidationErrors.Add(string.Format(SR.BindingMessageEncodingNotSupportedFormat, wsHttpBinding.MessageEncoding, WSMessageEncoding.Text));
                     }
+                    if (wsHttpBinding.Security.Mode == SecurityMode.Message)
+                    {
```
Comment (mconnew) [inline src/dotnet-svcutil/lib/src/CodeDomFixup/EndpointSelector.cs:66]: We don't have a second for it, but we also need to do this check for BasicHttpBinding, although there the mode type is BasicHttpSecurityMode. You need to remove the check for MessageEncoding just above this code as we now support that.

File: src/dotnet-svcutil/lib/src/Shared/TargetFrameworkHelper.cs [hunk 3]
```diff
@@ -63,6 +63,7 @@ internal class TargetFrameworkHelper
                 ProjectDependency.FromPackage("System.ServiceModel.Http", "4.8.*"    ),
                 ProjectDependency.FromPackage("System.ServiceModel.NetTcp", "4.8.*"  ),
                 ProjectDependency.FromPackage("System.ServiceModel.Security", "4.8.*"),
+                ProjectDependency.FromPackage("System.ServiceModel.Federation", "4.8.*"),
             } }
```
Comment (mconnew) [inline src/dotnet-svcutil/lib/src/Shared/TargetFrameworkHelper.cs:67]: We don't support Federation with svcutil yet so no need to add that package.

File: src/dotnet-svcutil/lib/src/CodeDomFixup/EndpointSelector.cs (hunk 2 above)
Comment (HongGit) [reply src/dotnet-svcutil/lib/src/CodeDomFixup/EndpointSelector.cs:66]: @ZhaodongTian  chatted with Matt, it's okay we are not supporting ReliableSession, and remove ling 61 to 64; and add check for BasicHttpBinding security mode as Matt suggested.

Thanks!

File: src/dotnet-svcutil/lib/src/Shared/TargetFrameworkHelper.cs (hunk 3 above)
Comment (ZhaodongTian) [reply src/dotnet-svcutil/lib/src/Shared/TargetFrameworkHelper.cs]: I have removed that package

File: src/dotnet-svcutil/lib/src/CodeDomFixup/EndpointSelector.cs (hunk 2 above)
Comment (ZhaodongTian) [reply src/dotnet-svcutil/lib/src/CodeDomFixup/EndpointSelector.cs:62]: Done, I have add the code for checking BasicHttpBinding security mode

**Review by HongGit — APPROVED** [review]

**Pull request merged**